File: Traditional-Important-Algorithms/exercises_py/04_os_algorithms/os09_reader_writer.py
# I AM NOT DONE

"""
os09_reader_writer.py

The Reader-Writer problem involves synchronizing access to a shared resource where
multiple readers can access the resource simultaneously, but writers need exclusive
access. This is a fundamental concurrency control problem.

Your task: Implement reader-writer locks with different fairness policies.
"""

import threading
from enum import Enum
from typing import Generic, TypeVar
import unittest
import time


T = TypeVar('T')


class LockPolicy(Enum):
    """Lock fairness policies."""
    READERS_PREFERENCE = "readers_preference"  # Readers are preferred, writers may starve
    WRITERS_PREFERENCE = "writers_preference"  # Writers are preferred, readers may starve
    FAIR = "fair"  # Fair scheduling, no starvation


class RwLock(Generic[T]):
    """Reader-writer lock with configurable fairness policy."""

    def __init__(self, data: T, policy: LockPolicy):
        self.data = data
        self.policy = policy
        self.lock = threading.Lock()
        self.read_condvar = threading.Condition(self.lock)
        self.write_condvar = threading.Condition(self.lock)
        self.readers = 0
        self.writers = 0
        self.waiting_readers = 0
        self.waiting_writers = 0

    def _can_read(self) -> bool:
        """
        TODO: Determine if a reader can acquire the lock based on policy.

        Returns:
            True if reader can acquire, False otherwise
        """
        pass  # TODO: Implement this

    def _can_write(self) -> bool:
        """
        TODO: Determine if a writer can acquire the lock.

        Writers need exclusive access (no readers, no other writers).

        Returns:
            True if writer can acquire, False otherwise
        """
        pass  # TODO: Implement this

    def read(self):
        """
        TODO: Acquire a read lock.

        The behavior depends on the policy:
        - ReadersPreference: Acquire if no writers
        - WritersPreference: Acquire if no writers and no waiting writers
        - Fair: Use a queue-like mechanism

        Returns:
            ReadGuard context manager
        """
        pass  # TODO: Implement this

    def write(self):
        """
        TODO: Acquire a write lock.

        Must wait until no readers and no other writers.

        Returns:
            WriteGuard context manager
        """
        pass  # TODO: Implement this

    def reader_count(self) -> int:
        """Get the number of active readers."""
        with self.lock:
            return self.readers

    def writer_count(self) -> int:
        """Get the number of active writers."""
        with self.lock:
            return self.writers

    def waiting_readers_count(self) -> int:
        """Get the number of waiting readers."""
        with self.lock:
            return self.waiting_readers

    def waiting_writers_count(self) -> int:
        """Get the number of waiting writers."""
        with self.lock:
            return self.waiting_writers


class ReadGuard(Generic[T]):
    """RAII-style guard for read access."""

    def __init__(self, rwlock: RwLock[T]):
        """
        TODO: Create a read guard.

        This should increment the reader count.

        Args:
            rwlock: The RwLock to guard
        """
        pass  # TODO: Implement this

    def __enter__(self):
        """Enter the context manager."""
        return self.data

    def __exit__(self, exc_type, exc_val, exc_tb):
        """
        TODO: Release the read lock.

        Decrement reader count and notify waiting writers.
        """
        pass  # TODO: Implement this


class WriteGuard(Generic[T]):
    """RAII-style guard for write access."""

    def __init__(self, rwlock: RwLock[T]):
        """
        TODO: Create a write guard.

        This should increment the writer count.

        Args:
            rwlock: The RwLock to guard
        """
        pass  # TODO: Implement this

    def __enter__(self):
        """Enter the context manager."""
        return self.data

    def __exit__(self, exc_type, exc_val, exc_tb):
        """
        TODO: Release the write lock.

        Decrement writer count and notify waiting readers/writers.
        """
        pass  # TODO: Implement this


class TestReaderWriter(unittest.TestCase):
    """Test cases for Reader-Writer Lock."""

    def test_single_reader(self):
        """Test single reader access."""
        lock = RwLock[int](42, LockPolicy.FAIR)
        guard = lock.read()
        with guard as data:
            self.assertEqual(data, 42)

    def test_single_writer(self):
        """Test single writer access."""
        lock = RwLock[int](42, LockPolicy.FAIR)
        guard = lock.write()
        with guard as data:
            data = 100  # Modify (in real implementation)
            self.assertEqual(data, 100)

    def test_multiple_readers(self):
        """Test multiple concurrent readers."""
        lock = RwLock[int](0, LockPolicy.READERS_PREFERENCE)
        results = []

        def reader():
            guard = lock.read()
            with guard as data:
                time.sleep(0.01)
                results.append(data)

        threads = []
        for _ in range(5):
            t = threading.Thread(target=reader)
            t.start()
            threads.append(t)

        for t in threads:
            t.join()

        # All readers should have run
        self.assertEqual(len(results), 5)

    def test_reader_writer_exclusion(self):
        """Test readers and writers are mutually exclusive."""
        lock = RwLock[int](0, LockPolicy.FAIR)
        results = []

        def writer():
            guard = lock.write()
            with guard as data:
                data = 1
                time.sleep(0.05)
                data = 2
                results.append(data)

        def reader():
            time.sleep(0.01)
            guard = lock.read()
            with guard as data:
                results.append(data)

        writer_thread = threading.Thread(target=writer)
        reader_thread = threading.Thread(target=reader)

        writer_thread.start()
        reader_thread.start()

        writer_thread.join()
        reader_thread.join()

        # Reader should see the final value after writer completes
        self.assertIn(2, results)

    def test_writers_preference(self):
        """Test writers preference policy."""
        lock = RwLock[int](0, LockPolicy.WRITERS_PREFERENCE)

        guard = lock.read()
        with guard:
            self.assertEqual(lock.reader_count(), 1)

        guard = lock.write()
        with guard:
            self.assertEqual(lock.writer_count(), 1)

        guard = lock.read()
        with guard:
            pass  # Should complete

    def test_readers_preference(self):
        """Test readers preference policy."""
        lock = RwLock[int](0, LockPolicy.READERS_PREFERENCE)

        def reader1():
            guard = lock.read()
            with guard:
                time.sleep(0.02)

        def writer():
            time.sleep(0.02)
            guard = lock.write()
            with guard:
                pass

        def reader2():
            time.sleep(0.01)
            guard = lock.read()
            with guard:
                pass

        r1 = threading.Thread(target=reader1)
        w = threading.Thread(target=writer)
        r2 = threading.Thread(target=reader2)

        r1.start()
        w.start()
        r2.start()

        r1.join()
        r2.join()
        w.join()

    def test_concurrent_readers(self):
        """Test multiple concurrent readers can access simultaneously."""
        lock = RwLock[list](list(range(5)), LockPolicy.FAIR)
        results = []

        def reader():
            guard = lock.read()
            with guard as data:
                total = sum(data)
                results.append(total)

        threads = []
        for _ in range(10):
            t = threading.Thread(target=reader)
            t.start()
            threads.append(t)

        for t in threads:
            t.join()

        # All readers should have seen the same data
        self.assertTrue(all(r == 10 for r in results))

    def test_sequential_writers(self):
        """Test writers execute sequentially."""
        lock = RwLock[int](0, LockPolicy.FAIR)
        results = []

        def writer(value):
            guard = lock.write()
            with guard:
                results.append(value)

        threads = []
        for i in range(5):
            t = threading.Thread(target=writer, args=(i,))
            t.start()
            threads.append(t)

        for t in threads:
            t.join()

        # All writers should have executed
        self.assertEqual(len(results), 5)
        self.assertEqual(set(results), {0, 1, 2, 3, 4})

    def test_alternating_read_write(self):
        """Test alternating read and write operations."""
        lock = RwLock[int](0, LockPolicy.FAIR)

        for i in range(10):
            if i % 2 == 0:
                guard = lock.write()
                with guard:
                    pass  # Write operation
            else:
                guard = lock.read()
                with guard:
                    pass  # Read operation

    def test_fair_policy(self):
        """Test fair policy prevents starvation."""
        lock = RwLock[int](0, LockPolicy.FAIR)

        def reader():
            guard = lock.read()
            with guard:
                time.sleep(0.01)

        def writer():
            time.sleep(0.01)
            guard = lock.write()
            with guard:
                pass

        # Hold read lock
        r = threading.Thread(target=reader)
        r.start()

        # Writer waits
        w = threading.Thread(target=writer)
        w.start()

        time.sleep(0.02)

        r.join()
        w.join()

    def test_no_writer_starvation(self):
        """Test writers eventually execute with writers preference."""
        lock = RwLock[int](0, LockPolicy.WRITERS_PREFERENCE)
        results = []

        def writer():
            time.sleep(0.01)
            guard = lock.write()
            with guard:
                results.append(999)

        def reader():
            guard = lock.read()
            with guard:
                time.sleep(0.005)

        threads = []

        # Spawn a writer that should eventually execute
        w = threading.Thread(target=writer)
        w.start()
        threads.append(w)

        # Spawn many readers
        for _ in range(5):
            r = threading.Thread(target=reader)
            r.start()
            threads.append(r)

        for t in threads:
            t.join()

        # Writer should have executed
        self.assertIn(999, results)

    def test_reader_count(self):
        """Test reader count tracking."""
        lock = RwLock[int](42, LockPolicy.FAIR)

        self.assertEqual(lock.reader_count(), 0)

        guard1 = lock.read()
        with guard1:
            self.assertEqual(lock.reader_count(), 1)

            guard2 = lock.read()
            with guard2:
                self.assertEqual(lock.reader_count(), 2)

            self.assertEqual(lock.reader_count(), 1)

        self.assertEqual(lock.reader_count(), 0)

    def test_writer_count(self):
        """Test writer count tracking."""
        lock = RwLock[int](42, LockPolicy.FAIR)

        self.assertEqual(lock.writer_count(), 0)

        guard = lock.write()
        with guard:
            self.assertEqual(lock.writer_count(), 1)

        self.assertEqual(lock.writer_count(), 0)

    def test_stress_test(self):
        """Test with many concurrent readers and writers."""
        lock = RwLock[int](0, LockPolicy.FAIR)
        num_iterations = 100

        def reader():
            for _ in range(num_iterations):
                guard = lock.read()
                with guard:
                    time.sleep(0.00001)

        def writer():
            for _ in range(num_iterations):
                guard = lock.write()
                with guard:
                    time.sleep(0.00001)

        threads = []

        # Spawn readers
        for _ in range(10):
            t = threading.Thread(target=reader)
            t.start()
            threads.append(t)

        # Spawn writers
        for _ in range(5):
            t = threading.Thread(target=writer)
            t.start()
            threads.append(t)

        for t in threads:
            t.join()


if __name__ == '__main__':
    unittest.main()

File: Traditional-Important-Algorithms/exercises/04_os_algorithms/os09_reader_writer.rs
// os09_reader_writer.rs
//
// The Reader-Writer problem involves synchronizing access to a shared resource where
// multiple readers can access the resource simultaneously, but writers need exclusive
// access. This is a fundamental concurrency control problem.
//
// Your task: Implement reader-writer locks with different fairness policies.

// I AM NOT DONE

use std::sync::{Arc, Condvar, Mutex};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LockPolicy {
    ReadersPreference,  // Readers are preferred, writers may starve
    WritersPreference,  // Writers are preferred, readers may starve
    Fair,               // Fair scheduling, no starvation
}

struct RwLockState {
    readers: usize,
    writers: usize,
    waiting_readers: usize,
    waiting_writers: usize,
}

pub struct RwLock<T> {
    data: Mutex<T>,
    state: Mutex<RwLockState>,
    read_condvar: Condvar,
    write_condvar: Condvar,
    policy: LockPolicy,
}

impl<T> RwLock<T> {
    pub fn new(data: T, policy: LockPolicy) -> Arc<Self> {
        // TODO: Create a new reader-writer lock with the given policy
        todo!()
    }

    pub fn read(&self) -> ReadGuard<T> {
        // TODO: Acquire a read lock
        // The behavior depends on the policy:
        // - ReadersPreference: Acquire if no writers
        // - WritersPreference: Acquire if no writers and no waiting writers
        // - Fair: Use a queue-like mechanism
        todo!()
    }

    pub fn write(&self) -> WriteGuard<T> {
        // TODO: Acquire a write lock
        // Must wait until no readers and no other writers
        todo!()
    }

    fn can_read(&self, state: &RwLockState) -> bool {
        // TODO: Determine if a reader can acquire the lock based on policy
        todo!()
    }

    fn can_write(&self, state: &RwLockState) -> bool {
        // TODO: Determine if a writer can acquire the lock
        // Writers need exclusive access (no readers, no other writers)
        todo!()
    }

    pub fn reader_count(&self) -> usize {
        let state = self.state.lock().unwrap();
        state.readers
    }

    pub fn writer_count(&self) -> usize {
        let state = self.state.lock().unwrap();
        state.writers
    }

    pub fn waiting_readers(&self) -> usize {
        let state = self.state.lock().unwrap();
        state.waiting_readers
    }

    pub fn waiting_writers(&self) -> usize {
        let state = self.state.lock().unwrap();
        state.waiting_writers
    }
}

pub struct ReadGuard<'a, T> {
    lock: &'a RwLock<T>,
    data: *const T,
}

impl<'a, T> ReadGuard<'a, T> {
    fn new(lock: &'a RwLock<T>) -> Self {
        // TODO: Create a read guard
        // This should increment the reader count
        todo!()
    }
}

impl<'a, T> Drop for ReadGuard<'a, T> {
    fn drop(&mut self) {
        // TODO: Release the read lock
        // Decrement reader count and notify waiting writers
        todo!()
    }
}

impl<'a, T> std::ops::Deref for ReadGuard<'a, T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        unsafe { &*self.data }
    }
}

pub struct WriteGuard<'a, T> {
    lock: &'a RwLock<T>,
    data: *mut T,
}

impl<'a, T> WriteGuard<'a, T> {
    fn new(lock: &'a RwLock<T>) -> Self {
        // TODO: Create a write guard
        // This should increment the writer count
        todo!()
    }
}

impl<'a, T> Drop for WriteGuard<'a, T> {
    fn drop(&mut self) {
        // TODO: Release the write lock
        // Decrement writer count and notify waiting readers/writers
        todo!()
    }
}

impl<'a, T> std::ops::Deref for WriteGuard<'a, T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        unsafe { &*self.data }
    }
}

impl<'a, T> std::ops::DerefMut for WriteGuard<'a, T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        unsafe { &mut *self.data }
    }
}

// Safe wrappers that ensure Send + Sync
unsafe impl<T: Send> Send for RwLock<T> {}
unsafe impl<T: Send + Sync> Sync for RwLock<T> {}

#[cfg(test)]
mod tests {
    use super::*;
    use std::thread;
    use std::time::Duration;

    #[test]
    fn test_single_reader() {
        let lock = RwLock::new(42, LockPolicy::Fair);
        let guard = lock.read();
        assert_eq!(*guard, 42);
    }

    #[test]
    fn test_single_writer() {
        let lock = RwLock::new(42, LockPolicy::Fair);
        let mut guard = lock.write();
        *guard = 100;
        assert_eq!(*guard, 100);
    }

    #[test]
    fn test_multiple_readers() {
        let lock = RwLock::new(0, LockPolicy::ReadersPreference);

        let mut handles = vec![];
        for _ in 0..5 {
            let lock_clone = Arc::clone(&lock);
            let handle = thread::spawn(move || {
                let guard = lock_clone.read();
                thread::sleep(Duration::from_millis(10));
                *guard
            });
            handles.push(handle);
        }

        for handle in handles {
            handle.join().unwrap();
        }
    }

    #[test]
    fn test_reader_writer_exclusion() {
        let lock = RwLock::new(0, LockPolicy::Fair);
        let lock_clone = Arc::clone(&lock);

        let writer = thread::spawn(move || {
            let mut guard = lock_clone.write();
            *guard = 1;
            thread::sleep(Duration::from_millis(50));
            *guard = 2;
        });

        thread::sleep(Duration::from_millis(10));

        let reader = thread::spawn(move || {
            let guard = lock.read();
            *guard
        });

        writer.join().unwrap();
        let value = reader.join().unwrap();

        // Reader should see the final value after writer completes
        assert_eq!(value, 2);
    }

    #[test]
    fn test_writers_preference() {
        let lock = RwLock::new(0, LockPolicy::WritersPreference);

        let guard = lock.read();
        assert_eq!(lock.reader_count(), 1);
        drop(guard);

        let mut guard = lock.write();
        *guard = 42;
        drop(guard);

        let guard = lock.read();
        assert_eq!(*guard, 42);
    }

    #[test]
    fn test_readers_preference() {
        let lock = RwLock::new(0, LockPolicy::ReadersPreference);
        let lock_clone = Arc::clone(&lock);

        let _reader1 = lock.read();

        let writer = thread::spawn(move || {
            thread::sleep(Duration::from_millis(20));
            let mut guard = lock_clone.write();
            *guard = 100;
        });

        thread::sleep(Duration::from_millis(10));

        // With readers preference, this reader should be able to acquire
        let _reader2 = lock.read();
        assert_eq!(lock.reader_count(), 2);

        drop(_reader1);
        drop(_reader2);

        writer.join().unwrap();
    }

    #[test]
    fn test_concurrent_readers() {
        let lock = RwLock::new(vec![1, 2, 3, 4, 5], LockPolicy::Fair);

        let mut handles = vec![];
        for _ in 0..10 {
            let lock_clone = Arc::clone(&lock);
            let handle = thread::spawn(move || {
                let guard = lock_clone.read();
                let sum: i32 = guard.iter().sum();
                sum
            });
            handles.push(handle);
        }

        for handle in handles {
            let sum = handle.join().unwrap();
            assert_eq!(sum, 15);
        }
    }

    #[test]
    fn test_sequential_writers() {
        let lock = RwLock::new(0, LockPolicy::Fair);

        let mut handles = vec![];
        for i in 0..5 {
            let lock_clone = Arc::clone(&lock);
            let handle = thread::spawn(move || {
                let mut guard = lock_clone.write();
                *guard += i;
            });
            handles.push(handle);
        }

        for handle in handles {
            handle.join().unwrap();
        }

        let guard = lock.read();
        assert_eq!(*guard, 10); // 0 + 1 + 2 + 3 + 4
    }

    #[test]
    fn test_alternating_read_write() {
        let lock = RwLock::new(0, LockPolicy::Fair);

        for i in 0..10 {
            if i % 2 == 0 {
                let mut guard = lock.write();
                *guard += 1;
            } else {
                let guard = lock.read();
                assert!(*guard > 0);
            }
        }

        let guard = lock.read();
        assert_eq!(*guard, 5); // 5 writes
    }

    #[test]
    fn test_fair_policy() {
        let lock = RwLock::new(0, LockPolicy::Fair);
        let lock_clone = Arc::clone(&lock);

        // Hold read lock
        let reader = lock.read();

        // Writer waits
        let writer = thread::spawn(move || {
            thread::sleep(Duration::from_millis(10));
            let mut guard = lock_clone.write();
            *guard = 42;
        });

        thread::sleep(Duration::from_millis(20));
        drop(reader);

        writer.join().unwrap();

        let guard = lock.read();
        assert_eq!(*guard, 42);
    }

    #[test]
    fn test_no_writer_starvation() {
        let lock = RwLock::new(0, LockPolicy::WritersPreference);

        let mut handles = vec![];

        // Spawn a writer that should eventually execute
        let lock_writer = Arc::clone(&lock);
        let writer = thread::spawn(move || {
            thread::sleep(Duration::from_millis(10));
            let mut guard = lock_writer.write();
            *guard = 999;
        });

        // Spawn many readers
        for _ in 0..5 {
            let lock_reader = Arc::clone(&lock);
            let handle = thread::spawn(move || {
                let guard = lock_reader.read();
                thread::sleep(Duration::from_millis(5));
                *guard
            });
            handles.push(handle);
        }

        for handle in handles {
            handle.join().unwrap();
        }

        writer.join().unwrap();

        let guard = lock.read();
        assert_eq!(*guard, 999);
    }

    #[test]
    fn test_reader_count() {
        let lock = RwLock::new(42, LockPolicy::Fair);

        assert_eq!(lock.reader_count(), 0);

        let _g1 = lock.read();
        assert_eq!(lock.reader_count(), 1);

        let _g2 = lock.read();
        assert_eq!(lock.reader_count(), 2);

        drop(_g1);
        assert_eq!(lock.reader_count(), 1);

        drop(_g2);
        assert_eq!(lock.reader_count(), 0);
    }

    #[test]
    fn test_writer_count() {
        let lock = RwLock::new(42, LockPolicy::Fair);

        assert_eq!(lock.writer_count(), 0);

        let _guard = lock.write();
        assert_eq!(lock.writer_count(), 1);

        drop(_guard);
        assert_eq!(lock.writer_count(), 0);
    }

    #[test]
    fn test_stress_test() {
        let lock = RwLock::new(0, LockPolicy::Fair);
        let num_iterations = 100;

        let mut handles = vec![];

        // Spawn readers
        for _ in 0..10 {
            let lock_clone = Arc::clone(&lock);
            let handle = thread::spawn(move || {
                for _ in 0..num_iterations {
                    let _guard = lock_clone.read();
                    thread::sleep(Duration::from_micros(10));
                }
            });
            handles.push(handle);
        }

        // Spawn writers
        for _ in 0..5 {
            let lock_clone = Arc::clone(&lock);
            let handle = thread::spawn(move || {
                for _ in 0..num_iterations {
                    let mut guard = lock_clone.write();
                    *guard += 1;
                    thread::sleep(Duration::from_micros(10));
                }
            });
            handles.push(handle);
        }

        for handle in handles {
            handle.join().unwrap();
        }

        let guard = lock.read();
        assert_eq!(*guard, 500); // 5 writers * 100 iterations
    }
}
